bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java

```diff
@@ -54,135 +53,188 @@
+     */
     private CompositeModuleHandlerFactory compositeFactory;
 
+    /**
+     * {@link Map} holding all scheduled {@link Rule} re-initialization tasks. The relation is {@link Rule}'s UID to
+     * re-initialization task as a {@link Future} instance.
+     */
     private Map<String, Future> scheduleTasks = new HashMap<String, Future>(31);
 
+    /**
+     * Performs the {@link Rule} re-initialization tasks.
+     */
     private ScheduledExecutorService executor;
 
-    private Gson gson;
-
     /**
      * Constructor of {@link RuleEngine}. It initializes the logger and starts tracker for {@link ModuleHandlerFactory}
      * services.
-     *
-     * @param bc {@link BundleContext} used for tracker registration and rule engine logger creation.
-     * @throws InvalidSyntaxException
      */
     public RuleEngine() {
         this.rules = new HashMap<String, RuntimeRule>(20);
         this.contextMap = new HashMap<String, Map<String, Object>>();
         this.moduleHandlerFactories = new HashMap<String, ModuleHandlerFactory>(20);
     }
 
+    /**
+     * This method is used to create a {@link CompositeModuleHandlerFactory} that handles all composite
+     * {@link ModuleType}s. Called from DS to activate the rule engine component.
+     *
+     * @param contex a {@link BundleContext} used for {@link CompositeModuleHandlerFactory} creation.
+     */
+    @Activate
+    protected void activate(BundleContext context) {
+        compositeFactory = new CompositeModuleHandlerFactory(context, mtRegistry, this);
+    }
+
+    /**
+     * Bind the {@link ModuleTypeRegistry} service - called from DS.
+     *
+     * @param moduleTypeRegistry a {@link ModuleTypeRegistry} service.
+     */
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC)
     protected void setModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
-        if (moduleTypeRegistry == null) {
-            mtRegistry.removeRegistryChangeListener(this);
-            mtRegistry = null;
+        mtRegistry = moduleTypeRegistry;
+        mtRegistry.addRegistryChangeListener(this);
+    }
+
+    /**
+     * Unbind the {@link ModuleTypeRegistry} service - called from DS.
+     *
+     * @param moduleTypeRegistry a {@link ModuleTypeRegistry} service.
+     */
+    protected void unsetModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
+        mtRegistry.removeRegistryChangeListener(this);
+        mtRegistry = null;
+    }
+
+    /**
+     * Bind the {@link StatusInfoCallback} service - called from DS.
+     *
+     * @param statusInfoCallback a {@link StatusInfoCallback} service.
+     */
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC)
+    protected void setStatusInfoCallback(StatusInfoCallback statusInfoCallback) {
+        this.statusInfoCallback = statusInfoCallback;
+        if (statusInfoCallback instanceof RuleRegistryImpl) {
```

Comment:
I agree, though I'm trying to "hide" the cyclic dependency. I'm not sure what the suggestion is. Is it to:
- Make the RuleEngine dependent on RuleRegistry instead of StatusInfoCallback
- then in the setRuleRegistry method, cast it to RuleRegistryImpl and call a protected method to inject the RuleEngine 
- then in this protected method in RuleRegistryImpl, inject back the StatusInfoCallback in RuleEngine?

I don't like this either, but if you prefer it, I can make it. IMHO the problem comes from trying to hide this cyclic dependency between the RuleEngine and the RuleRegistry. I'm not sure why we can't simply express it in DS - when you have a static dependency in one direction, and a dynamic in the other, the cycle should not be a problem.

Reply:
Sorry, I had a typo above. I meant "correct place should be within the RuleRegistry...".

No, what you describe is not my suggestion. We should completely get rid off the cyclic dependency. Effectively, the RuleRegistry depends on the RuleEngine, but not the other way round.
The RuleEngine merely requires a mechanism to pass status updates, correct? Therefore, the RuleRegistry itself does not need to be a `StatusInfoCallback` instance. It can create such an instance (which is NOT a `RuleRegistry` at the same time) and pass it to the `ruleEngine` in `RuleRegistry.setRuleEngine` (which is called by DS).

Reply:
Err, I should have read all our discussions back in July - I just noticed that we are trying to have the dependencies the other way round, from RuleEngine to RuleRegistry only...

Does your code pick up the [suggestion of a RuleRegistryChangeListener](https://github.com/eclipse/smarthome/issues/3635#issuecomment-318335109)? Is this what the `StatusInfoCallback` is effectively trying to do?

Reply:
RuleRegistryChangeListener will be implemented as a separate class and instance of it will belongs to RuleEngine, because the RuleEngine is already a listener to the ModuleTypeRegistry and the RegistryChangeListener interface do not permit two registries.

Reply:
ok, understood.